core/src/main/java/kafka/server/share/SharePartition.java

```diff
@@ -2484,6 +2497,243 @@ private long startOffsetDuringInitialization(long partitionDataStartOffset) thro
+            if (floorOffset == null) {
+                log.debug("Fetched batch record {} not found for share partition: {}-{}", recordBatch, groupId,
+                    topicIdPartition);
+                throw new InvalidRecordStateException(
+                    "Batch record not found. The request batch offsets are not found in the cache.");
+            }
+            return cachedState.subMap(floorOffset.getKey(), true, recordBatch.lastOffset(), true);
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    // Visible for testing.
+    List<RecordBatch> fetchAbortedTransactionRecordBatches(
+        Iterable<? extends RecordBatch> batches,
+        List<FetchResponseData.AbortedTransaction> abortedTransactions
+    ) {
+        lock.writeLock().lock();
+        try {
+            PriorityQueue<FetchResponseData.AbortedTransaction> abortedTransactionsHeap = abortedTransactionsHeap(abortedTransactions);
+            Set<Long> abortedProducerIds = new HashSet<>();
+            List<RecordBatch> recordsToArchive = new ArrayList<>();
+
+            for (RecordBatch currentBatch : batches) {
+                if (currentBatch.hasProducerId()) {
+                    // remove from the aborted transactions queue, all aborted transactions which have begun
+                    // before the current batch's last offset and add the associated producerIds to the
+                    // aborted producer set
+                    if (abortedTransactionsHeap != null) {
+                        while (!abortedTransactionsHeap.isEmpty() && abortedTransactionsHeap.peek().firstOffset() <= currentBatch.lastOffset()) {
+                            FetchResponseData.AbortedTransaction abortedTransaction = abortedTransactionsHeap.poll();
+                            abortedProducerIds.add(abortedTransaction.producerId());
+                        }
+                    }
+                    long producerId = currentBatch.producerId();
+                    if (containsAbortMarker(currentBatch)) {
+                        abortedProducerIds.remove(producerId);
+                    } else if (isBatchAborted(currentBatch, abortedProducerIds)) {
+                        log.debug("Skipping aborted record batch for share partition: {}-{} with producerId {} and " +
+                            "offsets {} to {}", groupId, topicIdPartition, producerId, currentBatch.baseOffset(), currentBatch.lastOffset());
+                        recordsToArchive.add(currentBatch);
+                    }
+                }
+            }
+            return recordsToArchive;
+        } finally {
+            lock.writeLock().unlock();
+        }
+    }
+
+    private PriorityQueue<FetchResponseData.AbortedTransaction> abortedTransactionsHeap(List<FetchResponseData.AbortedTransaction> abortedTransactions) {
+        if (abortedTransactions == null || abortedTransactions.isEmpty())
```

Comment:
I think we know that neither of these conditions is true. I'd prefer to take advantage of that knowledge so that the caller does not need to handle the null return here with its own conditional logic.